### ISSUE TITLE
Add a proper way to recycle vector slots in the C API

### ISSUE THREAD

**Issue by adamhutchings** [jabacat/jml#41]
Right now, when a vector is disposed of through the C API, its slots in the `vectors` array just keeps wasting memory.
We want it to be the case that when a new vector is made, it takes up the first free spot in the array if there is any and only extends the array if it needs to.

**Comment by Sophon96** [comment]
Something we should consider is whether this is a good idea or not; recycling slots means that erroneous client programs can have harder-to-diagnose bugs. For example, suppose a program using the C API
1. Creates a `jml::Vector` with some parameters
2. Eventually deletes the previous `jml::Vector`, but erroneously holds onto the index (essentially a use-after-free). The slot is now empty, and the next `jml::Vector` will occupy it
3. Creates a new `jml::Vector` with different parameters. The old slot is now occupied by this new `jml::Vector`
4. Tries to modify the old, destroyed `jml::Vector` (UAF)

This will end up modifying the new `jml::Vector`, which may work without error with some methods but will inexplicably fail or fail with seemingly unrelated errors when using other operators.

**Comment by adamhutchings** [comment]
This is true, but is also essentially the C tradeoff between safety and efficiency. In my mind, this method offers the _possibility_ of UAF bugs, as opposed to the _certainty_ of excess memory usage by unused vectors with the non-free option. Thoughts?
